File: CodeChef/OCT20B/ADDSQURE.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long int
#define V vector
#define F(a,b,c) for (a=b;a<=c;++a)
#define I cin
#define O cout
#define all(x) x.begin(),x.end()
#define ff first
#define umap unordered_map
#define mp make_pair
#define found(m, x) (m.find(x)!=m.end())

int main() {
    I.sync_with_stdio(false);
    I.tie(0);

    ll W, H, N, M;
    cin >> W >> H >> N >> M;

    ll i, j, new_no_of_squares = 0, temp, y, initial_no_of_squares = 0;
    umap<ll, bool> m1, m2;
    V<ll> a(N), b(N);
    /*
        i, j - looping variable
        new_no_of_squares - Max no. of squares after adding a new line
        temp - stores 'new_no_of_squares' for every y, 0 <= y <= H
        y - looping variable
        initial_no_of_squares - No. of squares before adding the new line
        m1 - stores the unique difference of every 'X'
        m2 - stores the unique difference of every 'Y'
        a - stores the values of 'X'
        b - stores the values of 'Y'
    */

    F(i,0,N-1)
        I >> a[i];

    F(i,0,M-1)
        I >> b[i];

    // Sort 'a'
    sort(all(a));

    // Storing the unique differences for every pair of elements in 'a'
    F(i,0,N-1)
        F(j,i+1,N-1)
            m1.insert(mp(a[j] - a[i], true));

    // Sort 'b'
    sort(all(b));

    // Storing the unique differences for every pair of elements in 'b'
    F(i,0,M-1)
        F(j,i+1,M-1)
            m2.insert(mp(b[j] - b[i], true));

    // Finding the initial no. of squares
    for (auto it: m1)
        if (found(m2, it.ff))
            ++initial_no_of_squares;

    F(y,0,H)
    {
        temp = 0;
        F(i,0,M-1)
            if (found(m1, abs(b[i] - y)) and !found(m2, abs(b[i] - y)))
                ++temp;
        new_no_of_squares = max(new_no_of_squares, temp);
    }

    ll final_no_of_squares = initial_no_of_squares + new_no_of_squares;

    O << final_no_of_squares << '\n';

    return 0;
}

File: CodeChef/OCT20B/ADDSQURE.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
import java.util.Arrays;

class ADDSQURE {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        final int W, H, N, M;
        W = sc.nextInt();
        H = sc.nextInt();
        N = sc.nextInt();
        M = sc.nextInt();

        int[] a = new int[N], b = new int[M];

        HashMap<Integer, Boolean> X = new HashMap<>();
        HashMap<Integer, Boolean> Y = new HashMap<>();

        int i, j, new_no_of_squares = 0, temp, y, initial_no_of_squares = 0;

        for (i = 0; i < N; ++i)
            a[i] = sc.nextInt();
        
        for (i = 0; i < M; ++i)
            b[i] = sc.nextInt();
        
        sc.close();
        
        Arrays.sort(a);

        for (i = 0; i < N; ++i)
            for (j = i + 1; j < N; ++j)
                X.put(a[j] - a[i], true);
        
        Arrays.sort(b);

        for (i = 0; i < M; ++i)
            for (j = i + 1; j < M; ++j)
                Y.put(b[j] - b[i], true);

        for (Map.Entry<Integer, Boolean> it : X.entrySet())
            if (Y.containsKey(it.getKey()))
                ++initial_no_of_squares;
        
        for (y = 0; y <= H; ++y)
        {
            temp = 0;
            for (i = 0; i < M; ++i)
                if (X.containsKey(Math.abs(b[i] - y)) && !Y.containsKey(Math.abs(b[i] - y)))
                    ++temp;
            new_no_of_squares = Math.max(new_no_of_squares, temp);
        }

        int final_no_of_squares = initial_no_of_squares + new_no_of_squares;

        System.out.println(final_no_of_squares);
    }
}